forc-pkg/src/pkg.rs

```diff
@@ -1364,17 +1374,138 @@ pub fn fetch_git(fetch_id: u64, name: &str, pinned: &SourceGitPinned) -> Result<
             let _ = std::fs::remove_dir_all(&path);
         }
         std::fs::create_dir_all(&path)?;
-
+        let repo_path = repo.path();
         // Checkout HEAD to the target directory.
         let mut checkout = git2::build::CheckoutBuilder::new();
         checkout.force().target_dir(&path);
         repo.checkout_head(Some(&mut checkout))?;
+
+        // Copy current_dir to target
+        let copy_options = fs_extra::dir::CopyOptions::new();
+        fs_extra::copy_items(&[repo_path], &path, &copy_options)?;
```

Comment:
A few lines above, we checkout only the commit that we care about to the target `path`.

It looks like this new code copies *everything* within the repo into the same `path`, rather than only the desired commit.

It looks like the aim here is to provide access to the associated git repository as a sub-directory, so that this repo can be queried (via `git`) during future searches of the checkouts directory. Does this sound correct?

If that's the case, ideally we should not have both the checkout and the whole repo as a subdirectory. Repos can get large, e.g. the `std` checkout is only ~6 MB for the Sway repo, but some repos can contain other large assets like media, etc. We should do our best to keep these checkouts as small as possible, and especially avoid doubling the repo contents.

As an alternative to this copy, I wonder if we can include the `.git` directory along with the checkout somehow? Or alternatively, strip down this repo to the commit we need, and copy it *rather* than doing the checkout?

Another alternative might be, rather than cloning the whole repo, we could only store some tiny file that just contained the `SourceGit` that was used to produce the commit. I guess this would only work for git sources specified via tag or commit hash, and wouldn't work for branches as we couldn't know which is the latest commit :thinking:

Reply:
Hmm, that is correct. I just checked the difference between the old method and this approach, for fetching std there seems to be around 8 MB of difference and the size of the `.git` folder is about also 8MB. Thanks for the detailed possible workarounds but I think I am a little confused.

By adding the copy, aren't we just adding the `.git` directory (on top of the already present checkout, because before executing the checkout I believe there is only the `.git` directory present in the `tmp` folder)? If that is the case can we really avoid it? Because in offline mode if the user wanted to add a git package by providing a branch, we should be able to find the most recent one we have locally. If we got the `.git` folders intact in the _all_ checkouts, we can find the most recent one in a checkout that we did for a non-branch version of the same repo. 

For example, imagine we are adding an `std` dependency that points to this branch, and after some while, we remove that and point to the latest release implicitly. In the meantime, this branch is also updated. In offline mode, if we try to add a dependency that points to this branch, (i believe) the most recent commit that we could find from this branch would be coming from the second checkout (the one that we did for the implicit dep declaration). So to be able to find those, I thought we need the full `.git` dir. 

On the other hand, this whole `.git` and size stuff got me thinking and I realized if we got the `.git` dir we don't need to store anything else and expand to the full repo from there once we need it. 
> if that's the case, ideally we should not have both the checkout and the whole repo as a subdirectory

I am not sure if that is what you meant here 😄 

Testing this approach yields  a difference from 20MB(with this PR) to 8MB. Before this PR same checkout seems to take about 11mb on disk. I will look into this in more in detail after our wallet stuff but wanted to clarify a little as I am a little confused rn 😄